server/src/main/java/org/opensearch/ingest/AbstractBatchingProcessor.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.ingest;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+/**
+ * Abstract base class for batch processors.
+ *
+ * @opensearch.internal
+ */
+public abstract class AbstractBatchingProcessor extends AbstractProcessor {
+
+    public static final String BATCH_SIZE_FIELD = "batch_size";
+    private static final int DEFAULT_BATCH_SIZE = 1;
+    protected final int batchSize;
+
+    protected AbstractBatchingProcessor(String tag, String description, int batchSize) {
+        super(tag, description);
+        this.batchSize = batchSize;
+    }
+
+    /**
+     * Internal logic to process batched documents, must be implemented by concrete batch processors.
+     *
+     * @param ingestDocumentWrappers {@link List} of {@link IngestDocumentWrapper} to be processed.
+     * @param handler                {@link Consumer} to be called with the results of the processing.
+     */
+    public abstract void internalBatchExecute(
```

Comment:
Should this be protected?

Also, naming nitpick, but maybe `subBatchExecute` or something else to indicate that this operates on the smaller batches would be better.